### PR TITLE
Don't synthesize primary pointer button unless buttons = 0

### PR DESCRIPTION
**Pull request by moffatman**
Some embeddings might send pointer events with buttons = 0x2 (right-click) for touch or stylus device kinds. If the primary button is synthesized for those events, they won't behave properly. Without this change it's not possible to trigger "secondary tap" events on a TapGestureRecognizer while using a stylus or on an iPad trackpad, since that recognizer will receive buttons = 0x3. 

Related:
https://github.com/flutter/engine/pull/27019

### CODE REVIEW

**Comment by goderbauer** [comment]
@dkwingsmt Can you take a look at this?

File: packages/flutter/test/gestures/gesture_binding_test.dart [hunk 1]
```diff
@@ -175,15 +175,15 @@ void main() {
...
+  test('Should synthesize kPrimaryButton for touch when no button is set', () {
     final Offset location = const Offset(10.0, 10.0) * ui.window.devicePixelRatio;
     const PointerDeviceKind kind = PointerDeviceKind.touch;
     final ui.PointerDataPacket packet = ui.PointerDataPacket(
       data: <ui.PointerData>[
         ui.PointerData(change: ui.PointerChange.add, kind: kind, physicalX: location.dx, physicalY: location.dy),
         ui.PointerData(change: ui.PointerChange.hover, kind: kind, physicalX: location.dx, physicalY: location.dy),
         ui.PointerData(change: ui.PointerChange.down, kind: kind, physicalX: location.dx, physicalY: location.dy),
-        ui.PointerData(change: ui.PointerChange.move, kind: kind, physicalX: location.dx, physicalY: location.dy),
+        ui.PointerData(change: ui.PointerChange.move, buttons: kSecondaryButton, kind: kind, physicalX: location.dx, physicalY: location.dy),
```
Comment (dkwingsmt) [inline packages/flutter/test/gestures/gesture_binding_test.dart:186]: This does not feel like an event sequence that the embedding would send in real situation, empty buttons at down then secondary buttons at move. Can you make this a separate test case? Making it as close to what you capture on iPadOS as possible would be preferred.

**Review by dkwingsmt — CHANGES_REQUESTED** [review]
The change is legit, but I have question for the test case.

File: packages/flutter/test/gestures/gesture_binding_test.dart [hunk 2]
```diff
@@ -230,21 +258,21 @@ void main() {
       expect(events[2], isA<PointerDownEvent>());
       expect(events[2].buttons, equals(kPrimaryButton));
       expect(events[3], isA<PointerMoveEvent>());
-      expect(events[3].buttons, equals(kPrimaryButton | kSecondaryStylusButton));
+      expect(events[3].buttons, equals(kSecondaryStylusButton));
```
Comment (dkwingsmt) [inline packages/flutter/test/gestures/gesture_binding_test.dart:261]: This is interesting. So I was aware that the primary button is synthesized besides the secondary. I think the intention was that: This case indicates that the stylus first touches, then the user presses the button, in which case `kPrimaryButton | kSecondaryStylusButton` should be the correct result. Do you think we should somehow handle this case, or does it conflict with what you recorded on real device?

Comment (moffatman) [reply packages/flutter/test/gestures/gesture_binding_test.dart:261]: I think most styluses that have a right-click button such as Surface Pen don't let you change the button during a press. I have only been able to get it to work by holding the button before touching the screen. So maybe I should split this test case as well?

Changing the buttons during a pointer's contact should be allowed, but we should rely on the engine to send an event with the correct buttons set. The point of this synthesis is to ensure we never have buttons = 0, if we always add the primary button, right-clicking on widgets doesn't work at all with a stylus.

Comment (dkwingsmt) [reply packages/flutter/test/gestures/gesture_binding_test.dart:261]: Ok, makes sense.

Comment (moffatman) [reply packages/flutter/test/gestures/gesture_binding_test.dart:261]: So should I split up this case too? I have a feeling these cases were not originally meant to be real continuous sequences, for example, hover events do not make much sense for touch devices.

Comment (dkwingsmt) [reply packages/flutter/test/gestures/gesture_binding_test.dart:261]: No I was imagining. Let's keep this case until someone complains.

**Review by dkwingsmt — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
@moffatman Could you rebase this on top the latest master to get all the tests to pass? Thanks!

**Comment by moffatman** [comment]
@goderbauer Can you merge this now?

**Pull request merged**